crypto/asn1/asn1_par.c

```diff
@@ -224,244 +228,342 @@ static int asn1_parse_constructed_type(
+
+  if (len != 1) {
+    if (BIO_puts(bp, ":BAD BOOLEAN") <= 0) {
+      return 0;
     }
-    if (object_length > 0) {
-      BIO_printf(bp, ":%u", current_pos[0]);
+    *dump_as_hex = 1;
+  }
+  if (len > 0) {
+    if (BIO_printf(bp, ":%u", data[0]) <= 0) {
+      return 0;
     }
-  } else if (tag == V_ASN1_BMPSTRING) {
-    /* do the BMP thang */
-  } else if (tag == V_ASN1_OCTET_STRING) {
-    int i, printable = 1;
-
-    parse_pos = object_start;
-    octet_string =
-        d2i_ASN1_OCTET_STRING(NULL, &parse_pos, object_length + header_length);
-    if (octet_string != NULL && octet_string->length > 0) {
-      parse_pos = octet_string->data;
-      /*
-       * testing whether the octet string is printable
-       */
-      for (i = 0; i < octet_string->length; i++) {
-        if (((parse_pos[i] < ' ') && (parse_pos[i] != '\n') &&
-             (parse_pos[i] != '\r') && (parse_pos[i] != '\t')) ||
-            (parse_pos[i] > '~')) {
-          printable = 0;
-          break;
-        }
-      }
-      if (printable)
-      /* printable string */
-      {
-        if (BIO_write(bp, ":", 1) <= 0) {
-          goto end;
-        }
-        if (BIO_write(bp, (const char *)parse_pos, octet_string->length) <= 0) {
-          goto end;
-        }
-      } else if (!dump)
-      /*
-       * not printable => print octet string as hex dump
-       */
-      {
-        if (BIO_write(bp, "[HEX DUMP]:", 11) <= 0) {
-          goto end;
-        }
-        for (i = 0; i < octet_string->length; i++) {
-          if (BIO_printf(bp, "%02X", parse_pos[i]) <= 0) {
-            goto end;
-          }
-        }
-      } else
-      /* print the normal dump */
-      {
-        if (!newline_printed) {
-          if (BIO_write(bp, "\n", 1) <= 0) {
-            goto end;
-          }
-        }
-        if (BIO_dump_indent(bp, (const char *)parse_pos,
-                            ((dump == -1 || dump > octet_string->length)
-                                 ? octet_string->length
-                                 : dump),
-                            dump_indent) <= 0) {
-          goto end;
-        }
-        newline_printed = 1;
+  }
+  return 1;
+}
+
+static int asn1_parse_octet_string_type(BIO *bp, const uint8_t *data, long len,
+                                        int dump, int *newline_printed) {
+  GUARD_PTR(bp);
+  GUARD_PTR(data);
+  GUARD_PTR(newline_printed);
+
+  const unsigned char *parse_pos = data;
+  int return_code = 0;
+  int dump_indent = 6;
+
+  ASN1_OCTET_STRING *octet_string =
+      d2i_ASN1_OCTET_STRING(NULL, &parse_pos, len);
+  if (octet_string != NULL && octet_string->length > 0) {
+    int printable = 1;
+    parse_pos = octet_string->data;
+
+    // Test whether the octet string is printable
+    for (int i = 0; i < octet_string->length; i++) {
+      if (((parse_pos[i] < ' ') && (parse_pos[i] != '\n') &&
+           (parse_pos[i] != '\r') && (parse_pos[i] != '\t')) ||
+          (parse_pos[i] > '~')) {
```

Comment:
NP: We could comment that, for our purposes, we only consider ASCII strings to be printable.